Add float support to libsodium custom ops

yanndupis opened the pull request:
Hey @mortendahl, 

Here is what I have right now. It's working with floats. I will have to use generics or create new classes to make it work with `uint8` as well. 

I have created this PR in case you have time to provide preliminary feedback. However we probably prefer to merge libsodium-ops to master then merge this pr to master. 

Thank you!

mortendahl commented on primitives/tf_encrypted/primitives/sodium/cc/easy_box_kernels.cc:
```diff
@@ -159,15 +162,16 @@ class SodiumEasyBoxOpenDetached : public OpKernel {
 
     Tensor* plaintext_t;
     TensorShape plaintext_shape = ciphertext_t.shape();
+    plaintext_shape.RemoveLastDims(1);
```
would be good to put in some assertions in the form of `OP_REQUIRES(context, ...)` here; for instance that the last dimension equals `sizeof(float)`

mortendahl commented on primitives/tf_encrypted/primitives/sodium/cc/easy_box_kernels.cc:
```diff
@@ -159,15 +162,16 @@ class SodiumEasyBoxOpenDetached : public OpKernel {
     //
 
-    auto res = crypto_box_open_detached(plaintext, ciphertext, mac, ciphertext_t.shape().num_elements(), nonce, pk_sender, sk_receiver);
+    auto res = crypto_box_open_detached(plaintext, ciphertext, mac, 16, nonce, pk_sender, sk_receiver);
```
16..? :)

yanndupis replied:
oups, I had this fixed, forgot to push it

mortendahl commented on primitives/tf_encrypted/primitives/sodium/python/easy_box_test.py:
```diff
@@ -22,24 +22,24 @@ def test_seal(self):
         plaintext_recovered = easy_box.open_detached(ciphertext, mac, nonce, pk_s, sk_r)
 
-        assert plaintext_recovered.shape == plaintext.shape
+        # assert plaintext_recovered.shape == plaintext.shape
```
this should still hold

mortendahl commented on primitives/tf_encrypted/primitives/sodium/python/easy_box_test.py:
```diff
@@ -22,24 +22,24 @@ def test_seal(self):
         ciphertext, mac = easy_box.seal_detached(plaintext, nonce, pk_r, sk_s)
 
-        assert ciphertext.raw.shape == plaintext.shape
+        # assert ciphertext.raw.shape == plaintext.shape
```
should be `ciphertext.raw.shape == plaintext.shape + (4,)` instead

mortendahl commented on primitives/tf_encrypted/primitives/sodium/python/easy_box_test.py:
```diff
@@ -22,24 +22,24 @@ def test_seal(self):
         pk_r, _ = easy_box.gen_keypair()
 
-        plaintext = tf.constant([1, 2, 3, 4], shape=(2, 2), dtype=tf.uint8)
+        plaintext = tf.constant([1, 2, 3, 4], shape=(2, 2), dtype=tf.float32)
```
perhaps copy the tests instead so we continue to support both `tf.uint8` and `tf.float32`

mortendahl commented on primitives/tf_encrypted/primitives/sodium/cc/easy_box_kernels.cc:
```diff
@@ -10,9 +10,11 @@
 #include "tensorflow/core/framework/variant_tensor_data.h"
 
 #include "sodium.h"
+#include <iostream>
```
leftover?

mortendahl commented on primitives/tf_encrypted/primitives/sodium/cc/ops.cc:
```diff
@@ -16,7 +16,7 @@ REGISTER_OP("SodiumEasyBoxGenNonce")
 
 REGISTER_OP("SodiumEasyBoxSealDetached")
-    .Input("plaintext: uint8")
+    .Input("plaintext: float32")
```
check out https://github.com/tf-encrypted/tf-big/blob/master/tf_big/cc/kernels/big_kernels.cc#L31-L47 and https://github.com/tf-encrypted/tf-big/blob/master/tf_big/cc/kernels/big_kernels.cc#L303-L304 for how to support `uint8` and `float32` at the same time

yanndupis commented:
Hey @mortendahl ,

I think I have addressed most of your comments, except one. I wasn't able to add tensorshape checks in `MakeShape`:

```
inline TensorShape makeShapeSeal<float>(const Tensor& ciphertext, OpKernelContext* ctx){
  auto ciphertext_shape = ciphertext.shape();
  ciphertext_shape.AddDim(sizeof(float));
  int last_dim = ciphertext_shape.dims() - 1;
  OP_REQUIRES(context, ciphertext_shape.dim_size(last_dim) == sizeof(float),
      errors::Internal("Last dim of ciphertext_shape should match sizeof(float)"));
  return ciphertext_shape;
```

With this function, I will be getting: 
```
In file included from external/primitives/tf_encrypted/primitives/sodium/cc/easy_box_kernels.cc:13:
external/primitives/tf_encrypted/primitives/sodium/cc/easy_box_kernels_utils.h:27:15: error: use of undeclared identifier 'context'
  OP_REQUIRES(context, ciphertext_shape.dim_size(last_dim) == sizeof(float),
              ^
external/primitives/tf_encrypted/primitives/sodium/cc/easy_box_kernels_utils.h:27:15: error: use of undeclared identifier 'context'
external/primitives/tf_encrypted/primitives/sodium/cc/easy_box_kernels_utils.h:27:3: error: non-void function 'makeShapeSeal' should return a value [-Wreturn-type]
  OP_REQUIRES(context, ciphertext_shape.dim_size(last_dim) == sizeof(float),
  ^
bazel-out/darwin-opt/bin/external/local_config_tf/include/tensorflow/core/framework/op_kernel.h:1780:7: note: expanded from macro 'OP_REQUIRES'
      return;
```
I understand that the problem is that `makeShapeSeal` expect to return a TensorShape, when `OP_REQUIRES` return void. just not sure yet how to handle this situation. Any idea are welcomed :). thank you. 

Note I also add to add plaintext_type to `sodium_easy_box_open_detached` to handle uint8 and float32 otherwise the compiler was complaining that it wasn't an input parameter. 

Thank you!

mortendahl commented:
BTW, why not add support for `int32`, `uint32`, `uint64`, `int6`, `double` while you're at it? Given the current code it should be straight forward!

yanndupis commented:
`plaintext_type` as been renamed to `plaintext_dtype`. We will parametrize test and support to additional types in a separate PR. Thanks @mortendahl for your help and review!
